File: src/emitter/mermaid/collect-nodes-into.ts
import { VISUAL_ELEMENT_TYPE } from "../../visual-graph/visual-element-type.js";
import type { VisualElement } from "../../visual-graph/visual-element.js";
import type { VisualNode } from "../../visual-graph/visual-node.js";

export function collectNodesInto(
  elements: readonly VisualElement[],
  out: Map<string, VisualNode>,
): void {
  for (const e of elements) {
    if (e.type === VISUAL_ELEMENT_TYPE.Node) {
      out.set(e.id, e);
    } else {
      collectNodesInto(e.elements, out);
    }
  }
}

File: crates/unsnarl-emitter-mermaid/src/collect_nodes_into.rs
//! Walks an element tree and indexes every [`VisualNode`] by id
//! into the supplied map.
//!
//! Mirrors `ts/src/emitter/mermaid/collect-nodes-into.ts`.

use std::collections::HashMap;

use unsnarl_visual_graph::visual_element::VisualElement;
use unsnarl_visual_graph::visual_node::VisualNode;

pub fn collect_nodes_into<'a>(
    elements: &'a [VisualElement],
    out: &mut HashMap<String, &'a VisualNode>,
) {
    for e in elements {
        match e {
            VisualElement::Node(n) => {
                out.insert(n.id().to_string(), n);
            }
            VisualElement::Subgraph(s) => {
                collect_nodes_into(s.elements(), out);
            }
        }
    }
}
